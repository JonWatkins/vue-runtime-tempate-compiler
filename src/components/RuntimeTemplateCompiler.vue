<template>
  <component :is="compiled" v-if="compiled" />
</template>

<script>
import { compileToFunctions } from 'vue-template-compiler'
import { merge } from '../utils/helpers'
import config from '../utils/config'

export default {
  name: 'RuntimeTemplateCompiler',

  inheritAttrs: false,

  props: {
    parent: {
      type: Object,
      default: undefined
    },

    template: {
      type: String,
      default: config.defaultRuntimeTemplate
    },

    templateProps: {
      type: Object,
      default: () => ({})
    }
  },

  data() {
    return {
      compiled: null
    }
  },

  beforeMount() {
    const compiled = compileToFunctions(this.template)
    const parent = this.parent || this.$parent
    const data = parent.$data || {}
    const props = parent.$props || {}
    const options = parent.$options || {}

    const computed = options.computed || {}
    const components = options.component || {}
    const methods = options.methods || {}
    const filters = options.filters || {}

    this.compiled = Object.assign({}, compiled, {
      name: 'RuntimeComponent',
      data() {
        return merge(data, props)
      },
      filters,
      components,
      computed,
      methods
    })
  }
}
</script>
