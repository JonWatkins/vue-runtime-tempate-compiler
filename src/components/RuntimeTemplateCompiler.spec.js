'use strict'

import { shallowMount, mount } from '@vue/test-utils'
import RuntimeTemplateCompiler from './RuntimeTemplateCompiler.vue'
let wrapper

describe('components/RuntimeTemplateCompiler', () => {
  beforeEach(() => {
    wrapper = shallowMount(RuntimeTemplateCompiler)
  })

  it('should have a name', () => {
    expect(wrapper.name()).toBe('RuntimeTemplateCompiler')
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to render a HTML string as a template', () => {
    wrapper = mount({
      components: {
        RuntimeTemplateCompiler
      },
      data() {
        return {
          name: 'Bob',
          template: '<h1>Hello {{ name }}</h1>'
        }
      },
      template: `
        <runtime-template-compiler :template="template" />
      `
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
