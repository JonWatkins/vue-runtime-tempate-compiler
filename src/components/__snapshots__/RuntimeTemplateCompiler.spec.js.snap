// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RuntimeTemplateCompiler should be able to render a HTML string as a template 1`] = `<h1>Hello Bob</h1>`;

exports[`components/RuntimeTemplateCompiler should render correctly 1`] = `<runtimecomponent-stub></runtimecomponent-stub>`;
